smolyay/surrogate.py

```diff
@@ -0,0 +1,311 @@
+        self.domain = domain
+        self.grid_generator = grid_generator
+        self._coefficients = None
+        self._grid = None
+        self._points = None
+        self._data = None
+
+    @property
+    def coefficients(self):
+        """list: coefficients of surrogate."""
+        if self._coefficients is not None:
+            return self._coefficients.tolist()
+        else:
+            return None
+
+    @property
+    def data(self):
+        """list: data at sampling grid points."""
+        if self._data is not None:
+            return self._data.tolist()
+        else:
+            return None
+
+    @property
+    def dimension(self):
+        """Numbers of independent variables."""
+        return self._domain.shape[0]
+
+    @property
+    def domain(self):
+        """list: domain of the function to be approximated."""
+        if self.dimension > 1:
+            return self._domain.tolist()
+        else:
+            return self._domain[0].tolist()
+
+    @domain.setter
+    def domain(self, value):
+        value = numpy.array(value, ndmin=2)
+        if value.ndim != 2 or value.shape[1] != 2:
+            raise TypeError('Domain must be dim x 2')
+        self._domain = value
+        self._reset_grid()
+
+    @property
+    def grid(self):
+        """:class:`IndexGrid` Sampling grid."""
+        if self._grid is None:
+            self._grid = self.grid_generator(self.dimension)
+        return self._grid
+
+    @property
+    def grid_generator(self):
+        """:class:`IndexGridGenerator`: Sampling grid generator."""
+        return self._grid_generator
+
+    @grid_generator.setter
+    def grid_generator(self, value):
+        if not isinstance(value, IndexGridGenerator):
+            raise TypeError('Grids must be IndexGridGenerator')
+        self._grid_generator = value
+        self._reset_grid()
+
+    @property
+    def points(self):
+        """numpy.ndarray: Points to be sampled."""
+        if self._points is None:
+            self._points = self._mapdomain(
+                self.grid.points, [[-1, 1]]*self.dimension, self._domain)
+        return self._points
+
+    def __call__(self, x):
+        """Evaluate surrogate at a given input.
+
+        Parameter
+        ---------
+        x: list (for dimension > 1), or float (for dimension = 1)
+            Input.
+
+        Returns
+        -------
+        float
+            Surrogate output at x.
+
+        Raises
+        ------
+        ValueError
+            For surrogate to be evaluated, function needs to be trained.
+        IndexError
+            Input of the surrogate must be of length of the
+            function's dimensionality.
+        """
+        if self._coefficients is None:
+            raise ValueError('Function needs training!')
+        x = numpy.array(x, copy=False, ndmin=1)
+        if x.shape != (self.dimension, ):
+            raise IndexError('Input must match dimension of domain.')
```

Comment:
You should probably add an error check here that x lies within the domain:
```
if not (numpy.all(x >= self._domain[:,0]) and numpy.all(x <= self._domain[:,1])):
    raise ValueError('x must lie in domain of surrogate')
```
Then add a test for this too. I think we might have discussed how the basis functions _should_ catch this themselves, but having the error raised here will make more sense to the caller.